ambry-mysql/src/main/java/com/github/ambry/mysql/BatchUpdater.java

```diff
@@ -0,0 +1,153 @@
+  // The time when the first statement is added
+  private long startTime = 0;
+
+  /**
+   * Constructor to instantiate a {@link BatchUpdater}.
+   * @param dataAccessor The {@link MySqlDataAccessor}.
+   * @param sql The sql command to prepare the statement.
+   * @param tableName The name of the table that update statements are targeting at.
+   * @param maxBatchSize The max batch size.
+   * @throws SQLException
+   */
+  public BatchUpdater(MySqlDataAccessor dataAccessor, String sql, String tableName, int maxBatchSize)
+      throws SQLException {
+    this.dataAccessor = Objects.requireNonNull(dataAccessor, "MySqlDataAccessor is empty");
+    this.tableName = Objects.requireNonNull(tableName, "TableName is empty");
+    if (maxBatchSize < 0) {
+      throw new IllegalArgumentException("MaxBatchSize is not valid: " + maxBatchSize);
+    }
+    this.maxBatchSize = maxBatchSize;
+    try {
+      autoCommit = dataAccessor.getAutoCommmit();
+      dataAccessor.setAutoCommit(false);
+      statement = dataAccessor.getPreparedStatement(sql, true);
+      statement.clearBatch();
+    } catch (SQLException e) {
+      dataAccessor.onException(e, Write);
+      logger.error("Failed to prepare for batch insert on {}", tableName, e);
+      throw e;
+    }
+  }
+
+  /**
+   * Add statement to the batch. Since the sql statement provided to the constructor is a prepared statement,
+   * this method only takes a lambda to supply values to the {@link PreparedStatement}.
+   * @param valueSupplier The consumer to supply the values. The {@link PreparedStatement} is the parameter to this
+   *                      consumer. It can also throws an {@link SQLException}.
+   * @throws SQLException
+   */
+  protected synchronized void addUpdateToBatch(GenericThrowableConsumer<PreparedStatement, SQLException> valueSupplier)
+      throws SQLException {
+    try {
+      if (startTime == 0) {
+        startTime = System.currentTimeMillis();
+      }
+      if (maxBatchSize != 0 && currentBatchSize >= maxBatchSize) {
+        executeBatch();
+      }
+      valueSupplier.accept(statement);
+      statement.addBatch();
+      currentBatchSize++;
+      totalBatchSize++;
+    } catch (SQLException e) {
+      dataAccessor.onException(e, Write);
+      logger.error("Failed to add batch on {}", tableName, e);
+      throw e;
+    }
+  }
+
+  /**
+   * Commit all the sql commands added to this batch. If commit fails, the batch would be rolled back.
+   * @throws SQLException
+   */
+  public void commit() throws SQLException {
+    try {
+      executeBatch();
+      dataAccessor.commit();
+      if (startTime != 0) {
+        dataAccessor.onSuccess(Write, System.currentTimeMillis() - startTime);
+      }
+    } catch (SQLException e) {
+      dataAccessor.onException(e, Write);
+      logger.error("Failed to commit batch on {}, rolling back, batch size {}", tableName, totalBatchSize, e);
+      throw e;
+    } finally {
+      try {
+        dataAccessor.rollback();
```

Comment:
Huh?  Why are you calling rollback in the finally block?  This should only happen in the exception path, no?

Reply:
typo here, updated.